Invalid generator for VS2019 x64

In CMakeSettings.json, Visual Studio still uses the old style of specifying the generator architecture "Visual Studio 16 2019 Win64", while CMake does not support this style anymore and instead requests using "Visual Studio 16 2019" and additionally the parameter "-A Win64". 

The task runcmake seems to use the exact generator as written in the CMakeSettings.json, but it should detect VS2019 and higher and extract the architecture from the generator field. Otherwise the build on azure breaks with the error message:

```
Processing configuration: 'x64-Release'. 
  | :\Program Files\CMake\bin\cmake.exe" -G "Visual Studio 16 2019 Win64" -DCMAKE_BUILD_TYPE=RelWithDebInfo  d:\a\1\s |  
  | CMake Error: Could not create named generator Visual Studio 16 2019 Win64 |  
  |   |  
  | Generators |  
  | * Visual Studio 16 2019         = Generates Visual Studio 2019 project files. |  
  |                                   Use -A option to specify architecture.
```

--- a/README.md
+++ b/README.md
@@ -93,11 +93,15 @@ To package the extension for release purpose, run:
 
 or to pack it for development purpose:
 
+  > export MAJOR=3
+  > export MINOR=0
   > npm run packdev
 
-In the latter case GUIDs of extension and tasks will be different.
-and the name of the task will have appended "-dev".
-  
+`MAJOR` and `MINOR` environment variables are used to set versions of extension and tasks. Note that the path is bumped automatically.
+With 'packdev' GUIDs of extension and tasks are changed so the extension could be
+uploaded and tested on Azure DevOps without interfering with already released tasks.
+The name of the tasks have appended "-dev" to distringuish them.
+
 ### Testing
 
 Run the whole test suite:

--- a/task-cmake/src/cmakesettings-runner.ts
+++ b/task-cmake/src/cmakesettings-runner.ts
@@ -216,6 +216,11 @@ export class CMakeSettingsJsonRunner {
   doBuild: boolean;
   ninjaPath: string;
   sourceScript: string;
+  static readonly ARM64: [string, string] = ["ARM64", "ARM64"];
+  static readonly ARM: [string, string] = ["ARM", "ARM"];
+  static readonly X64: [string, string] = ["x64", "x64"];
+  static readonly WIN64: [string, string] = ["Win64", "x64"];
+  static readonly WIN32: [string, string] = ["Win32", "Win32"];
 
   constructor(settingsPath: any, configurationFilter: string, buildArgs: string, root: string, vcpkgTriplet: string, useVcpkgToolchain: boolean, doBuild: boolean, ninjaPath: string, sourceScript: string) {
     this.cmakeSettingsJson = settingsPath;
@@ -257,6 +262,43 @@ export class CMakeSettingsJsonRunner {
 
     return globalEnvs;
   }
+
+  getGeneratorArgs(gen: string): string {
+    if (gen.includes("Visual Studio")) {
+      // for VS generators, add the -A value
+      let architectureParam: string | undefined = undefined;
+      let generatorParam: string = "Visual Studio 16 2019";
+
+      const architectures: [string, string][] = [
+        CMakeSettingsJsonRunner.X64,
+        CMakeSettingsJsonRunner.WIN32,
+        CMakeSettingsJsonRunner.WIN64,
+        CMakeSettingsJsonRunner.ARM64, // Note ARM64 must be replaced before ARM!
+        CMakeSettingsJsonRunner.ARM
+      ];
+
+      // Remove the platform 
+      for (let architecture of architectures) {
+        if (gen.includes(architecture[0])) {
+          gen = gen.replace(architecture[0], "");
+          architectureParam = architecture[1];
+        }
+      }
+
+      gen = `-G \"${gen.trim()}\"`;
+
+      if (architectureParam) {
+        gen += ` -A ${architectureParam}`
+      }
+    }
+    else {
+      // All non-VS generators are passed as is.
+      gen = `-G "${gen}"`;
+    }
+
+    return gen;
+  }
+
 
   async run(): Promise<void> {
     let content: any = fs.readFileSync(this.cmakeSettingsJson);
@@ -306,7 +348,8 @@ export class CMakeSettingsJsonRunner {
       // The build directory goes into the artifact directory.
       configuration.buildDir = path.join(utils.getArtifactsDir(), configuration.name);
 
-      cmakeArgs += ` -G "${configuration.generator}"`;
+      cmakeArgs += this.getGeneratorArgs(configuration.generator);
+
       if (utils.isNinjaGenerator(cmakeArgs)) {
         let ninjaPath: string = ninjalib.retrieveNinjaPath(this.ninjaPath);
         cmakeArgs += ` -DCMAKE_MAKE_PROGRAM="${ninjaPath}"`;

--- a/task-cmake/tests/_suite.ts
+++ b/task-cmake/tests/_suite.ts
@@ -125,6 +125,23 @@ describe('CMake task tests', function () {
       });
     });
 
+    it('cmakesettings.json with VS generators: provide proper -G -A values to cmake',
+    (done: MochaDone) => {
+      testutils.runTest(done, (done) => {
+        let tp = path.join(__dirname, basePath, 'success-cmakesettings-vsgenerators.js');
+        let tr: ttm.MockTestRunner = new ttm.MockTestRunner(tp);
+        tr.run();
+        outputStdout(tr.stdout);
+        assert.equal(tr.succeeded, true, 'should have succeeded');
+        assert.equal(tr.warningIssues.length, 0, 'should have no warnings');
+        assert.equal(tr.errorIssues.length, 0, 'should have no errors');
+        assert.equal(
+          tr.stdout.indexOf('anyName') >= 0, true,
+          'should contain "anyName"');
+      });
+    });
+
+
   it('cmakesettings.json with BOM and comments should succeed',
     (done: MochaDone) => {
       testutils.runTest(done, (done) => {
@@ -154,9 +171,12 @@ describe('CMake task tests', function () {
         assert.equal(tr.errorIssues.length, 0, 'should have no errors');
         assert(tr.stdOutContained('x86-Debug'), 'should contain "x86-Debug"');
         assert(
-          tr.stdOutContained('-G "Visual Studio 15 2017"'),
-          'should contain "-G "Visual Studio 15 2017"');
-        assert(
+          tr.stdOutContained('-G "Visual Studio 16 2019"'),
+          'should contain "-G "Visual Studio 16 2019"');
+          assert(
+            tr.stdOutContained('-A x64'),
+            'should contain "-A x64');
+          assert(
           tr.stdOutContained('-DCMAKE_BUILD_TYPE="Debug"'),
           'should contain configuration type Debug');
         // Check for the presence of one and only one toolchain passed only.

--- a/task-cmake/tests/success-cmakesettings-complex.ts
+++ b/task-cmake/tests/success-cmakesettings-complex.ts
@@ -17,7 +17,7 @@ let answers: ma.TaskLibAnswers = <ma.TaskLibAnswers>{
   'checkPath': { '/usr/local/bin/cmake': true, '/usr/local/bin/node': true },
   'exec': {
     '/usr/local/bin/cmake': { 'code': 0, 'stdout': 'cmake test output here' },
-    '/usr/local/bin/cmake -G Visual Studio 15 2017 -DCMAKE_BUILD_TYPE=Debug -DCMAKE_TOOLCHAIN_FILE:string=D:/src/vcpkg/scripts/buildsystems/vcpkg.cmake .':
+    '/usr/local/bin/cmake -G Visual Studio 16 2019 -A x64 -DCMAKE_BUILD_TYPE=Debug -DCMAKE_TOOLCHAIN_FILE:string=D:/src/vcpkg/scripts/buildsystems/vcpkg.cmake .':
       { 'code': 0, 'stdout': 'cmake output here' },
     '/usr/local/bin/cmake -G Ninja -DCMAKE_BUILD_TYPE=Debug -DCMAKE_ANY_FLAG:boolean=OFF -DCONFIGURATION:STRING=ResolvedConfiguration .': {
       'code': 0, 'stdout': 'cmake output here'
@@ -59,7 +59,7 @@ tmr.registerMock('fs', {
             }
           ],
           'name': 'x86-Debug',
-          'generator': 'Visual Studio 15 2017',
+          'generator': 'Visual Studio 16 2019 x64',
           'configurationType': '${CONFIGURATIONTYPE}',
           'inheritEnvironments': ["ENVNAME"],
           'buildRoot':

--- a/task-cmake/tests/success-cmakesettings-vsgenerators.ts
+++ b/task-cmake/tests/success-cmakesettings-vsgenerators.ts
@@ -1,0 +1,79 @@
+import * as ma from 'azure-pipelines-task-lib/mock-answer';
+import * as tmrm from 'azure-pipelines-task-lib/mock-run';
+import * as path from 'path';
+
+import { Globals } from '../src/globals'
+
+import * as utils from './test-utils'
+
+let taskPath = path.join(__dirname, '..', 'src', 'cmake-task.js');
+let tmr: tmrm.TaskMockRunner = new tmrm.TaskMockRunner(taskPath);
+
+// Arrange
+let answers: ma.TaskLibAnswers = <ma.TaskLibAnswers>{
+  'which': { 'cmake': '/usr/local/bin/cmake', 'node': '/usr/local/bin/node' },
+  'checkPath': { '/usr/local/bin/cmake': true, '/usr/local/bin/node': true },
+  'exec': {
+    '/usr/local/bin/cmake': { 'code': 0, 'stdout': 'cmake test output here' },
+    '/usr/local/bin/cmake -G Visual Studio 16 2019 -A ARM64 -DCMAKE_BUILD_TYPE=RelWithDebInfo .':
+      { 'code': 0, 'stdout': 'this is the cmake output' },
+    '/usr/local/bin/cmake --build . -- -cmake -build -args':
+      { 'code': 0, 'stdout': 'cmake build output here' }
+  },
+  'mkdirP': { '/': { 'code': 0, 'stdout': 'mkdirP output' } }
+};
+
+class MockStats {
+  mode = 600;
+};
+tmr.registerMock('fs', {
+  writeFileSync: function (filePath, contents) { },
+  existsSync: function (filePath, contents) {
+    return true;
+  },
+  readFileSync: function (filePath) {
+    return '{\
+      //comment\n\
+      "configurations": [{\
+      "environments": [{ "envvar": "envvalue"}],\
+        "name": "anyName",\
+        // I love comments\n\
+        "configurationType": "RelWithDebInfo",\
+        "generator": "Visual Studio 16 2019 ARM64",\
+        "buildRoot": "/build/root/${env.envvar}",\
+        "buildCommandArgs": "-cmake -build -args"\
+      }]\
+    }';
+  },
+  statSync: function (filePath) {
+    let s: MockStats = new MockStats();
+    return s;
+  },
+  chmodSync: function (filePath: string) { }
+});
+
+tmr.registerMock('./utils', {
+  getSourceDir: function (): string {
+    return '/agent/w/1/s';
+  },
+  isWin32: function (): boolean {
+    return true;
+  },
+  injectEnvVariables: function (a, b): void { },
+  getArtifactsDir: function (): string { return '/agent/w/1/a'; },
+  build: function (): void { },
+  injectVcpkgToolchain: function (args: string, triplet: string): string { return args; },
+  isNinjaGenerator: function (): boolean { return false; }
+});
+
+tmr.setAnswers(answers);
+utils.clearInputs();
+tmr.setInput(Globals.cmakeListsOrSettingsJson, 'CMakeSettingsJson');
+tmr.setInput(Globals.cmakeSettingsJsonPath, 'anyCMakeSettings.json');
+tmr.setInput(Globals.configurationRegexFilter, 'any.+');
+tmr.setInput(Globals.buildWithCMake, 'true');
+tmr.setInput(Globals.buildWithCMakeArgs, 'this must be unused');
+process.env["BUILD_ARTIFACTSTAGINGDIRECTORY"] = "/agent/w/1/a/";
+
+// Act
+tmr.run();

